### PR TITLE
Remove dangerous stringref::c_str()

### PR DESCRIPTION
**Pull request by vekterli**
@baldersheim please review.
@toregge @arnej27959 FYI, feel free to have a look as well.

The expected semantics of `c_str()` (a null-terminated string) cannot
be satisfied with a string reference, so remove the function entirely
to prevent people from using it in buggy ways.

Replaces `c_str()` with `data()` in places where it is presumed safe,
otherwise constructs temporary string instances. Certain call-sites
have been de-stringref'd in favor of regular strings, in particular
where C APIs have been transitively called. The vast majority of
these were called with string parameters anyway, so should not
cause much extra allocation.

### CODE REVIEW

**Comment by vekterli** [comment]
It may also be mentioned that the vast majority of our faffing around with `c_str()` could be removed entirely by moving logging etc. to a type-aware/type _safe_ formatting library like [fmtlib](http://fmtlib.net/latest/index.html). We could then just happily pass strings/stringrefs/iostreamable objects without the need for any temporary allocs.

File: document/src/vespa/document/base/field.cpp [hunk 1]
```diff
@@ -91,15 +91,15 @@ Field::validateId(int newId) {
         throw vespalib::IllegalArgumentException(vespalib::make_string(
                     "Attempt to set the id of %s to %d failed, values from "
                     "100 to 127 are reserved for internal use",
-                    getName().c_str(), newId));
+                    vespalib::string(getName()).c_str(), newId));
     }
```
Comment (baldersheim) [inline document/src/vespa/document/base/field.cpp]: Field:.getName() return ref to zero terminated string.

**Review by baldersheim — APPROVED** [review]
Looks good, only thing I see is that it is not neccessary to use vespalib::string() around FieldBase.getName() stringrefs. I checked both implementations and they both return zero terminated strings.

Merge at will

File: document/src/vespa/document/base/field.cpp (hunk 1 above)
Comment (vekterli) [reply document/src/vespa/document/base/field.cpp]: Good catch, will remove wrapping for these

**Pull request merged**